### PR TITLE
Fixes #1482, removed unsafe double RLock

### LINKED ISSUE #1482: Impossible to stop Snap task
We have a task that runs for approximately 2 minutes and then is stopped. The task is being run in a loop of configurable number of iterations. Sometimes it is not possible to stop the task.

- OS version
````
[root@*********** ~]# uname -a
Linux ******* 4.9.2-1.el7.elrepo.x86_64 #1 SMP Tue Jan 10 11:30:19 EST 2017 x86_64 x86_64 x86_64 GNU/Linux
````
- Snap version
````
[root@*********** ~]# snapteld --version
snapteld version 1.0.0
````
- Environment details (virtual, physical, etc.) - bare metal
- Steps to reproduce - keep creating and stopping Snap task.
- Expected results: - task stoppped
- Actual results: task is not stopped and pretends to be running; one of the plugins remains running despite of attempt to kill it
````
[root@*********** ~]# snaptel task list
ID 					 NAME 						 STATE 		 HIT 	 MISS 	 FAIL 	 CREATED 		 LAST FAILURE
dc2810f5-76e2-4c91-a4b4-08918d55df8f 	 serenity                                  	 Running 	 434 	 26 	 1 	 1:17PM 1-20-2017 	 CollectMetrics call error : could not obtain load metric: no history stored: could not calculate current load
````

````
[root@*********** ~]# ps ax | grep [s]nap
16600 ?        Ssl    0:13 /opt/snap/sbin/snapteld
32595 ?        Sl     0:01 /tmp/478065240/snap-plugin-publisher-influxdb {"LogLevel":5,"PingTimeoutDuration":10000000000,"NoDaemon":false,"Pprof":false}
````

@marcin-krolik suggested that it might be related to https://github.com/intelsdi-x/snap/issues/1454.

Log file is available at: https://gist.github.com/iwankgb/03ca6ac4d18cd1247fd22905645b3dbb

### PR DESCRIPTION
**Pull request by katarzyna-z**
Fixes #1482

Summary of changes:
- removed unsafe double RLock, Rlock is hold before `SelectAP()` in these lines: [L404](https://github.com/intelsdi-x/snap/blob/master/control/available_plugin.go#L404), [L453](https://github.com/intelsdi-x/snap/blob/master/control/available_plugin.go#L453), [L487](https://github.com/intelsdi-x/snap/blob/master/control/available_plugin.go#L487)

Testing done:
- tested using this script: [test.sh](https://gist.github.com/katarzyna-z/0d1fb8c3bfad134e494bd57390293f89)

@intelsdi-x/snap-maintainers

### CODE REVIEW

**Comment by IRCody** [comment]
What makes the double RLock unsafe and how is it related to #1482?

**Comment by katarzyna-z** [comment]
in golang documentation (https://golang.org/src/sync/rwmutex.go) there is following information:
```golang
// If a goroutine holds a RWMutex for reading, it must not expect this or any
// other goroutine to be able to also take the read lock until the first read
// lock is released. In particular, this prohibits recursive read locking.
// This is to ensure that the lock eventually becomes available;
// a blocked Lock call excludes new readers from acquiring the lock.
```
I've also found this issue: https://github.com/golang/go/issues/15418.
Testing Snap I've noticed that there is a problem with locking mutex in (*pool).Unsubscribe and (*pool).Subscribe.

File: control/strategy/pool.go [hunk 1]
```diff
@@ -361,10 +361,8 @@ func (p *pool) SubscriptionCount() int {
 }
 
 // SelectAP selects an available plugin from the pool
+// the method is not thread safety, it should be protected outside of the body
```
Comment (jcooklin) [inline control/strategy/pool.go]: s/safety/safe/

**Comment by IRCody** [comment]
Thanks for the links. That's an interesting and subtle difference (more subtle if you read the entire comment on the type).

Do we want the locks to be removed from inside pool or for them to be used inside pool only? Other pool functions contain the same code for locking/unlocking that we are removing here, do they have a similar bug?

File: control/strategy/pool.go (hunk 1 above)
Comment (katarzyna-z) [inline control/strategy/pool.go]: Corrected :]

**Comment by katarzyna-z** [comment]
I've found one more double RLock in `(p *pool) Eligible()` (there are Rlocks in `(p *pool) Count()` and `(p *pool) SubscriptionCount()`) and lock for reading in `(p *pool) IncRestartCount` where value is modified.

I anlyzed locking of pool and pool is locked outside only three times in:
`(ap *availablePlugins) collectMetrics`, `(ap *availablePlugins) publishMetrics`, `(ap *availablePlugins) processMetrics` so it seems that the rest of locks inside pool is correct.

**Comment by IRCody** [comment]
The changes in this PR do seem to fix some pretty serious issues that are identified by the script. The failures we were seeing were a fail by the tester(me) and not the PR 👎 . With fresh eyes the fix works as described/expected.

I think we might want to have more discussion about whether we want the pool mutex to be locked from inside or outside the pool. I think it might be confusing to do both (since that is at least partially what led to this issue). I know after talking with @jcooklin about it that we want to have some consistency. Maybe we can have that discussion here?

**Comment by katarzyna-z** [comment]
I've updated the PR - I moved Rlocks outside pool to `(p *pool) SelectAP`,  please see [4cf884c](https://github.com/intelsdi-x/snap/pull/1525/commits/4cf884c386e9bc3189f26b38f1726770756bb1a4). Testing Snap I don't see bad consequences of that change but I would rather remove the latest commit in my PR. 
Taking look at code [control/available_plugin.go#L443](https://github.com/intelsdi-x/snap/blob/master/control/available_plugin.go#L443), before moving the Rlock to  `(p *pool) SelectAP` pool was locked for reading (L453) till L471. When RLock is moved to `(p *pool) SelectAP` scope of lock is shorter, there are operation on `availablePlugin` but `available plugin` may be removed by `(p *pool) SelectAndKill`([control/strategy/pool.go#L307](https://github.com/intelsdi-x/snap/blob/master/control/strategy/pool.go#L307) and [control/strategy/pool.go#L331](https://github.com/intelsdi-x/snap/blob/master/control/strategy/pool.go#L331)).

**Comment by jcooklin** [comment]
>Testing Snap I don't see bad consequences of these changes but I would rather remove the latest commit in my PR....

**Comment by katarzyna-z** [comment]
I've remove the latest commit and squashed the rest of commits.

@IRCody @jcooklin Thanks for tests and review! :rose:

**Comment by PatrykMatyjasek** [comment]
LGTM

**Pull request merged**